### ISSUE TITLE
Change div tags to script tags and put viewContext/props inside

### ISSUE THREAD

**Issue by martyphee** [shakacode/react_on_rails#412]
> this is exacerbated because HTML encoded JSON is very bloated as `"` -> `&quot` when rendered into an HTML attribute and there are a great deal of quote marks in JSON!

On pages with large amounts of data the time to usability is dramatically increased on lower bandwidth connections as megabytes of html encoded JSON needs to be downloaded before the HTML that displays the page can be retrieved.

**Comment by martyphee** [comment]
I'm working on a PR right now to address this.

**Comment by justin808** [comment]
Turbo links is the tough one if we do that. Script tag is how I originally had it. Other issue is escaping the data.

**Comment by martyphee** [comment]
What issues does this cause with TurboLinks?

If the data is put into the script tag as proper JSON what issue's might we still have?

**Comment by justin808** [comment]
The page scripts don't run with TurboLinks, without some special configuration, and I'm not even sure on that.

So TurboLinks will just insert the HTML tags on the existing page (so CSS and JS are not reloaded), and we have a hook from TurboLinks to parse the special dom elements and kick start the React parts. See https://github.com/turbolinks/turbolinks#working-with-script-elements

**Comment by martyphee** [comment]
According to the link

> Turbolinks evaluates <script> elements in a page’s <body> each time it renders the page.


Looking at the link you provided the `<script/>` takes would be re-evaluated.  It would only contain the props and no JS code.

**Comment by justin808** [comment]
Let's try it out! It's possible that the older turbo links didn't.

**Comment by martyphee** [comment]
I'll try to do some work on this over the weekend.  I'm headed back to
London Sunday and probably won't have much time during the week, but I'll
try to fit some work in on it.

**Comment by justin808** [comment]
@martyphee Any info on this? Be particularly careful with escaping issues. I.e., suppose you have application data sent in the JavasScript, and it can close the script tag and open a new one.

**Comment by martyphee** [comment]
I'll try to take a look at this today.

**Comment by martyphee** [comment]
I got some work done on it, but haven't finished.

**Comment by jbhatab** [comment]
@martyphee sounds like a great improvement. Is the difference noticeable on phone load time?

**Comment by martyphee** [comment]
@jbhatab I have most of the test passing now.  I'll hopefully get the rest done in today or tomorrow.  Mainly the ones which just check the rendering of the `div`'s.

**Comment by martyphee** [comment]
Ok, down to two test failures.

**Comment by justin808** [comment]
@martyphee The most important thing is that we super, super carefully consider the escaping and that we have good tests. If we incorrectly put the data in the generated JavaScript, this would allow for script injection hack.

**Comment by martyphee** [comment]
Won't that be more of the server making sure it doesn't except or removes all XSS attempts?  Not sure the best way of escaping the data on the server without causing other side effects.

**Comment by squadette** [comment]
@justin808: just in case: http://apidock.com/rails/ERB/Util/json_escape properly escapes JSON-encoded strings for further inclusion to HTML code (inside the tags).

It handles things like `{ "foo": "<script>window.alert(1);</script>"}`.  It does JSON-level escaping, so that HTML-dangerous chars do not leak to the HTML code. 

You can have a code like this:

````
  def json_safe_and_pretty(something)
    if Rails.env.development?
      json_escape(JSON.pretty_generate(something.as_json))
    else
      json_escape(something.to_json)
    end
  end
````

and use it in your ERB template like this:

````
<script type="application/json" id="json-entries"><%= raw json_safe_and_pretty(@foobar) %></script>
````

**Comment by gregoryStarr** [comment]
Hey Guys, I have a project I would love to use this on but we have a hard requirement of not being able to mount our state into the hidden divs the way this project does.. Has there been any progress on this or is there any way I can jump in and give a hand?

**Comment by justin808** [comment]
@gregoryStarr it would be awesome to get your help. Both @martyphee and I would appreciate it. 

Please see the full message thread above. The main hazard is an XSS injection in the data.

**Comment by justin808** [comment]
@gregoryStarr Please work off of #411.

**Comment by justin808** [comment]
Closed by #775.
